### DISCUSSION TITLE
Ask user for overwriting when in interactive mode

### DISCUSSION THREAD

**Discussion by abravalheri** [pyscaffold/pyscaffold#392] [Ideas]
In future versions, how about we ask the user if we find an existing file marked as `no_overwrite` during updates in interactive mode? (if `--force` is passed all the files should be silently overwritten)

We could have a `Existing file: <PATH>, overwrite? ([Y]es, [n]o, [d]iff)` message (stdlib's `difflib.unified_diff` could be used)...

Is it worthy to add this to PyScaffold, when the flow with `git difftool` works fine? (Sometimes I get a bit tired of manually reverting all the files when I just want to have the up-to-date version of a few ones)

An alternative also could be having a `--overwrite` flag, marking files as safe for updates...

**Comment by abravalheri** [comment]
Related but independent discussion that might be worthy its own topic or not (if we have lots of discussion, I might move it):

`no_overwrite` is by far the most common use case I have found in all the extensions we have and in the core... Should it be the default behaviour in v5 ou v6 and a `safe_overwrite` be created to explicitly indicate when the file can be overwritten?

It is a breaking change, but would it be worthy the hassle?

**Comment by FlorianWilhelm** [comment]
Interesting topic which surely depends a lot on personal taste. Due to my FOMA, I actually tend to let PyScaffold always overwrite everything. [BTW: We do have a check that the repo is clean before doing so, right?] This allows me to see what I would get and then I use some `git difftool` like Meld or Fork to combine my old configuration with the new one. This works for me quite fine and is highly flexible, also I am not missing out on new features, which would have happened if my configs were not overridden. 

Thus my take on the first question would be that in many cases it's not new vs old but rather some merge between new and old based on what the user wants. For this task, external tools are much more powerful, and thus adding `difflib` to PyScaffold might not be so useful in the long run. But maybe extending the Update section in the docs and explaining best practices when updating might be pretty helpful for a lot of users.

Regarding the second question, I agree that changing the default might help to simplify things. Would it then be 'overwrite' instead of 'safe_overwrite'? Why safe? I think it would be worth the hassle if it simplifies things. Which files do you think `update` would then change? Only `setup.py` and maybe `conf.py`? Hm... some more radical idea. What if an update just reads in the PyScaffold section and overwrites the whole structure telling the user in the end to use `git difftool` to decide what he/she wants? This would simplify a lot but also make the update process more cumbersome for some users...

**Comment by abravalheri** [comment]
Yes, we do the check, if the latest changes are not committed we abort the update 😄 

Currently, the files that are always updated are:
- The `__init__.py` files
- `docs/{conf.py,readme.rst,license.rst,changelog.rst,Makefile,requirements.txt}`
   - maybe we should make `docs/requirements.txt` a `no_overwrite`  😅
   - `docs/conf.py` is likely to be in the category of files that users always want to merge to get the latest nice stuff
- `setup.py` (⇒ probably always update unless someone is using Cython or similar)

I agree that `git difftool` is the best tool for the job and that dealing with `difflib` might be overkill. I suppose my workflow is very similar to yours, I start with a `putup -U -f` and then I go diffing/merging... What usually happens is that I want to have the latest updates (merge) of pre-commit, tox, conf.py, but then I mostly certainly don't like to have README.rst, docs/index.rst, setup.cfg, CHANGELOG.rst and tests/conftest.py being messed up... With `--force` is always all or nothing, so I always end up reverting the same files all over with `git checkout HEAD -- file1 file2 ...`. That is why I though that a system where you can mark files would be good (with either a `--overwrite` for opting specific files in or `--no-overwrite` for opting specific files out).

Regarding names for a possible change in v5/6, I don't mind just `overwrite`. I though in `safe_overwrite`, but to be more clear it should be `safe_for_overwrite` to convey the idea that the file is likely to not have user customization...

I like the idea of the more aggressive update... Maybe it would work fine with the opt-in/opt-out flags (and ultimately the users could keep these allow/deny lists in the `$HOME/.config/pyscaffold/default.cfg` file as a preference).

**Reply by abravalheri** [reply]
In terms of crazy, complicated and unnecessary alternative ways, what is also possible is keeping a SHA1 of all the files stored somewhere when the user creates the project and then comparing during updates, always overwriting the ones that were never changed (a.k.a `pyscaffold.lock`)

**Reply by FlorianWilhelm** [reply]
Hm... actually this could be done *quite easily*. When running `putup` one could just save the sha1 of the initial commit (done by PyScaffold) in setup.cfg. During an update one could then ask git if there was any change to the file. It might become difficult if you do an update after an update though. In this case one would need to somehow get the sha1 of the commit right after `putup --update` is run or one would have to somehow store the sha1 before always.

**Reply by abravalheri** [reply]
It is a bit of egg-chicken problem isn't it? PyScaffold does create the initial commit, but then we need to edit the `setup.cfg` file to add the sha1, which makes the repository dirty and needing a commit...

I wonder if it would be easy to add a "mark" in the commit message and then have PyScaffold to `git log --grep` it to find the commit SHA...

But that would still not solve the fact that `--update --force` is all or nothing and to have the latest updates of `pre-commit.yaml` I have to overwrite my `CHANGELOG.rst`... (Sorry just thinking out loud on the internet)

**Reply by FlorianWilhelm** [reply]
I think it would work if one would always safe the current commit sha1 during an update. Since `--update` checks that the environment is clean the sha1 always referes to a valid state. For the initial setup there is of course no current sha1. In this case, one would need a special sha1 to reflect that, like `before-root-commit` as [root commit seems to be the official name for the initial commit](https://raw.githubusercontent.com/git/git/master/Documentation/RelNotes/1.7.12.txt).

**Reply by abravalheri** [reply]
Oh I see. That is clever! I suppose this way we could do a `git rev-list $sha..HEAD --ancestry-path --first-parent` and get the first sha on the list (unfortunately --max-count and --reverse are a bit counter-intuitive in `git rev-list`)

So what do you think, of the 2 options:

1. a) Always store the SHA of the previous commit (before init/updates), b) `--update` will overwrite everything that was not changed since SHA+1, user can opt-in for overwriting more files with `--overwrite FILE`    **OR**
2. a) Go more aggressive than currently we are (we can prevent changing things like CHANGELOG) and then have a `--no-overwrite FILE`.
(both `--overwrite` and `--no-overwrite` being also readable as preferences from `~/.config/pyscaffold/default.cfg`

**Reply by FlorianWilhelm** [reply]
So 2. would mean that I have to define all files I don't want to be overwritten with the `--no-overwrite` flag or in a config file? Not sure that would be so user-friendly. Most users I guess wouldn't know about that flag and everything would be overwritten anyway.

 Option 1 seems like quite a bit of implementation and it might have some unwanted side effects.

Maybe we should first start to think about what kind of files we have. We have
1. *essential files* for the packaging like pyproject.toml, setup.cfg, setup.py
2. *config files* for other tooling like pre-commit, tox, conf.py, etc.
3.  *doc files* like Sphinx rst, CHANGELOG, README etc.

Maybe doc files should never be overwritten by default if they were changed by the users. Changes are mostly large I guess, if you for instance have your own `index.rst`. Config files could be overwritten by default as using `git difftool` to reapply the minimal changes most users do (at least that's what I guess) should be really easy. Essential files need to be consistent so they need to be overwritten or at least some smart update protocol needs to be applied like for `setup.cfg` if we would need for instance some other configuration.

**Reply by abravalheri** [reply]
That is food for thought, thank you very much @FlorianWilhelm.
Let's start by making `no_overwrite` the default in version 6... if we manage to find an easy/simple implementation for the git sha1 trick we can than try to make the overwrite smarter considering the 3 categories that you have mentioned.

I will open an issue just for tracking the milestone, since I don't have any intention to work with this right now :P

Meanwhile, I will make `docs/requirements.txt` a no overwrite for v5, because that can be considered kind of a bug, right?

**Reply by abravalheri** [reply]
Just some useful snippets in the context of the git SHA trick:

- SHA of the root commit in the repo: `git rev-list --all --max-parents=0 --first-parent `
  - `git log --grep 'Initial commit' --format='%H'` can also be used as a fallback
- List of commits between `$sha` and `HEAD`: `git rev-list $sha..HEAD --ancestry-path --first-parent`
- List of files that *have changed* compared to`$sha`: `git diff --name-only $sha`

Pseudo-code for finding all the files that *have changed* since the last update:

```python
def get_root_commit():
    try:
        return sh("git rev-list --all --max-parents=0 --first-parent")
    except ShellCommandException:
        return sh("git log --grep 'Initial commit' --format='%H'")

def following_commit(ref):
    return sh(f"git rev-list {ref}..HEAD --ancestry-path --first-parent").splitlines()[-1].strip()
    

ref= setupcfg['pyscaffold']['last_update_reference']  # any better name for this field?
last_update = get_root_commit() if ref == 'before-root-commit' else following_commit(ref)
changed_files = [f.strip() for f in sh(f"git diff --name-only {ref}").splitlines()]
```

**Reply by julie777** [reply]
> It is a bit of egg-chicken problem isn't it? PyScaffold does create the initial commit, but then we need to edit the `setup.cfg` file to add the sha1, which makes the repository dirty and needing a commit...
> 
Actually, we could separate git init from the initial commit and then changes to setup.cfg would be part  of the initial commit.

**Comment by julie777** [comment]
My thoughts

1. we should never overwrite anything automatically - which means I agree with the no overwrite default
2. it would be nice to be able to know whether the src or the dest changed before overwriting. This will be more important after implementing #608 
3. it is also easy to check if overwriting is actually going to change the file before worrying about the question

I haven't started working with updating yet. Some of the overwrite or not could also be handled by showing a diff and and asking the user what they want to do for that file.
